pkg/preflight/validate.go

```diff
@@ -0,0 +1,401 @@
+	tmpPVCSpec := buildTmpPVC(pvc, dstSC)
+	tmpPVC, err := client.CoreV1().PersistentVolumeClaims(tmpPVCSpec.Namespace).Create(
+		ctx, tmpPVCSpec, metav1.CreateOptions{})
+	if err != nil {
+		if !k8serrors.IsAlreadyExists(err) {
+			return nil, fmt.Errorf("failed to create temporary pvc %s: %w", tmpPVCSpec.Name, err)
+		}
+
+		// PVC exist, get it
+		tmpPVC, err = client.CoreV1().PersistentVolumeClaims(tmpPVCSpec.Namespace).Get(ctx, tmpPVCSpec.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get existing temp PVC %s: %w", tmpPVCSpec.Name, err)
+		}
+	}
+
+	// consume pvc to determine any access mode errors
+	pvcConsumerPodSpec := buildTmpPVCConsumerPod(tmpPVC.Name, tmpPVC.Namespace, tmpPodImage)
+	pvcConsumerPod, err := client.CoreV1().Pods(tmpPVC.Namespace).Create(ctx, pvcConsumerPodSpec, metav1.CreateOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create validation pod %s: %w", pvcConsumerPodSpec.Name, err)
+	}
+
+	// cleanup pvc and pod at the end
+	defer func() {
+		// pod must be deleted first then the pvc
+		if err = deletePVConsumerPod(client, pvcConsumerPod); err != nil {
+			l.Printf("failed to cleanup pv consumer pod %s: %s", pvcConsumerPod.Name, err)
+		}
+
+		if err = deleteTmpPVC(l, client, tmpPVC); err != nil {
+			l.Printf("failed to cleanup tmp claim: %s", err)
+		}
+	}()
+
+	podReadyTimeoutEnd := time.Now().Add(timeout)
+	for {
+		gotPod, err := client.CoreV1().Pods(pvcConsumerPodSpec.Namespace).Get(ctx, pvcConsumerPodSpec.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get pv consumer pod %s: %w", gotPod.Name, err)
+		}
+
+		switch {
+		case k8spodutils.IsPodReady(gotPod):
+			return nil, nil
+		default:
+			time.Sleep(time.Second)
+		}
+
+		if time.Now().After(podReadyTimeoutEnd) {
+			// The volume consumer pod never went into running phase which means it's probably an error
+			// with provisioning the volume.
+			// A pod in Pending phase means the API Server has created the resource and stored it in etcd,
+			// but the pod has not been scheduled yet, nor have container images been pulled from the registry.
+			if gotPod.Status.Phase == corev1.PodPending {
+				// check pvc status and get error
+				pvcPendingError, err := getPVCError(client, tmpPVC)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get pvc failure: %w", err)
+				}
+				return pvcPendingError, nil
+			}
+			// pod failed for other reason(s)
+			return nil, fmt.Errorf("unexpected status for pod %s: %s", gotPod.Name, gotPod.Status.Phase)
+		}
+	}
+}
+
+// deleteTmpPVC deletes the provided pvc from the default namespace and waits until the
+// backing pv dissapear as well (this is mandatory so we don't leave any orphan pv as this would
+// cause pvmigrate to fail). this function has a timeout of 5 minutes, after that an error is
+// returned.
+func deleteTmpPVC(l *log.Logger, client k8sclient.Interface, pvc *corev1.PersistentVolumeClaim) error {
```

Comment:
Couldn't this function be simplified  ? I mean, we are deleting just one pvc so we don't need the `waitFor` loop, right ? In this scenario I think that it could be something like:

1. find the pv related to the pvc (that would be just one).
2. delete pvc.
3. wait until the one pv found on step 1 dissapear.

I may have misunderstood this so feel free to ignore this comment.